packages/SwingSet/src/kernel/kernel.js

```diff
@@ -913,7 +913,13 @@ export default function buildKernel(
       return results;
     }
 
-    const args = [upgradeID, true, undefined, incarnationNumber];
+    const kernelRootObjSlot = exportRootObject(kernelKeeper, vatID);
+    const args = [
+      upgradeID,
+      true,
+      undefined,
+      { incarnationNumber, rootObject: kslot(kernelRootObjSlot) },
```

Comment:
Hm.

You're right that there's nothing fundamentally wrong about re-supplying the root object in the result value of the upgrade request: anybody who had access to the control facet would have been given the root object the first time around, and it's unlikely that someone would go and split those authorities ("I'm going to let you upgrade this vat but *not* access the root object"??).

My hesitation is that re-supplying it lends the impression that each version has its own root, whereas really each version has its own *implementation* of a shared root-object identity. Re-supplying it raises the possibility that the root returned by `upgradeVat` might be different than the one you got from `createVat`. Not re-supplying it removes that concept entirely.

OTOH, I can imagine an upgrade scenario in which the original caller failed to retain the root object (maybe they used it to build a secondary object and then dropped it), and now they wish they'd kept it around (perhaps to make another secondary object). Doing a null upgrade with this re-supply feature would let them recover from that mistake.

So.. ok, let's leave it in. But let's add a note to the `upgradeVat` documentation to make it clear that this is a second copy of the same original object, and that it's not like `upgradeVat` is creating a new identity.

Reply:
Actually @erights do you have any opinion on this? Should having the control facet for a vat (which allows `terminateVat` and `upgradeVat`) also give you access to the root object for that vat? Can you imagine a scenario in which it would be surprising for `upgradeVat()` (which previously only returned the new incarnation number, powerless) to lead to that root object?